table partition: add RANGE COLUMNS partitioning (#10385)
pingcap/docs: partitioned-table.md

* You want to use a column that contains time or date values, or containing values arising from some other series.
* You need to frequently run queries on the columns used for partitioning. For example, when executing a query like `EXPLAIN SELECT COUNT(*) FROM employees WHERE separated BETWEEN '2000-01-01' AND '2000-12-31' GROUP BY store_id;`, TiDB can quickly know that only the data in the `p2` partition needs to be scanned, because the other partitions do not match the `WHERE` condition.

### Range COLUMNS partitioning

> Range COLUMNS partitioning is a variant of Range partitioning. You can use one or more columns as partitioning key and the partition columns data type can be integer, string (`char`/`varchar`), `date` and `datetime`. You cannot use any expressions, like non-COLUMNS partitioning, only the columns directly.
> 
> Suppose you want to partition by name as well as be able to easily drop older no-longer valid data, then you can create a table like this:

Reviewer: ```suggestion
Range COLUMNS partitioning is a variant of Range partitioning. You can use one or more columns as partitioning keys. The partition columns data types can be integer, string (`char`/`varchar`), `date` and `datetime`. Any expressions, such as non-COLUMNS partitioning, are not supported.

Suppose that you want to partition by name, and drop old and invalid data, then you can create a table as follows:

```

```sql
CREATE TABLE t (
  valid_until datetime,
  name varchar(255) CHARACTER SET ascii,
  notes text
)
PARTITION BY RANGE COLUMNS(name,valid_until)
(PARTITION `p2022-g` VALUES LESS THAN ('G','2023-01-01 00:00:00'),
 PARTITION `p2023-g` VALUES LESS THAN ('G','2024-01-01 00:00:00'),
 PARTITION `p2024-g` VALUES LESS THAN ('G','2025-01-01 00:00:00'),
 PARTITION `p2022-m` VALUES LESS THAN ('M','2023-01-01 00:00:00'),
 PARTITION `p2023-m` VALUES LESS THAN ('M','2024-01-01 00:00:00'),
 PARTITION `p2024-m` VALUES LESS THAN ('M','2025-01-01 00:00:00'),
 PARTITION `p2022-s` VALUES LESS THAN ('S','2023-01-01 00:00:00'),
 PARTITION `p2023-s` VALUES LESS THAN ('S','2024-01-01 00:00:00'),
 PARTITION `p2024-s` VALUES LESS THAN ('S','2025-01-01 00:00:00'),
 PARTITION `p2022-` VALUES LESS THAN (0x7f,'2023-01-01 00:00:00'),
 PARTITION `p2023-` VALUES LESS THAN (0x7f,'2024-01-01 00:00:00'),
 PARTITION `p2024-` VALUES LESS THAN (0x7f,'2025-01-01 00:00:00'))
```

> Which will split the data per year and names in the ranges [''-'G'), ['G'-'M'), ['M'-'S'), ['S',) allowing to easily drop unvalid data while still benefit from partition pruning on both name and valid_until columns.

Reviewer: "allowing to easily drop invalid data while still benefit from partition pruning on both name and valid_until columns."
This sentence is confusing. Could you refine it? @mjonss

> Which will split the data per year and names in the ranges [''-'G'), ['G'-'M'), ['M'-'S'), ['S',) allowing to easily drop unvalid data while still benefit from partition pruning on both name and valid_until columns.

Reviewer: ```suggestion
It will split the data by year and by name in the ranges [''-'G'), ['G'-'M'), ['M'-'S'), ['S',), allowing you to easily drop invalid data while still benefit from partition pruning on both `name` and `valid_until` columns.
```
